crates/runtime/src/dataconnector/github/github_app_token_provider.rs

```diff
@@ -0,0 +1,339 @@
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use async_trait::async_trait;
+use chrono::{DateTime, Utc};
+use data_components::token_provider::{Error, Result, TokenProvider};
+use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
+use serde::{Deserialize, Serialize};
+use snafu::prelude::*;
+use tokio::sync::RwLock;
+
+#[derive(Debug, Snafu)]
+pub enum GitHubAppError {
+    #[snafu(display("Invalid private key"))]
+    InvalidPrivateKey { source: jsonwebtoken::errors::Error },
+
+    #[snafu(display("Unable to get system time"))]
+    UnableToGetSystemTime { source: std::time::SystemTimeError },
+
+    #[snafu(display("Invalid system time"))]
+    InvalidSystemTime { source: std::num::TryFromIntError },
+
+    #[snafu(display("Unable to generate JWT"))]
+    UnableToGenerateJWT { source: jsonwebtoken::errors::Error },
+
+    #[snafu(display("Unable to get GitHub installation access token"))]
+    UnableToGetGitHubInstallationAccessToken { source: reqwest::Error },
+
+    #[snafu(display("Unable to get GitHub installation access token body"))]
+    UnableToGetGitHubInstallationAccessTokenBody { source: reqwest::Error },
+}
+
+#[derive(Serialize)]
+struct Claims {
+    iat: usize,
+    exp: usize,
+    iss: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct TokenResponse {
+    token: String,
+    expires_at: String,
+}
+
+#[async_trait]
+trait TokenGenerator: Send + Sync {
+    async fn generate_token(
+        &self,
+        app_client_id: Arc<str>,
+        private_key: Arc<str>,
+        installation_id: Arc<str>,
+    ) -> Result<TokenResponse, GitHubAppError>;
+}
```

Comment:
I don't think this needs to be a trait? Do we expect multiple instances?

Reply:
No, but I needed to separate that into separate struct for isolated tests.